mixing/signatures.go

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2023-2024 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package mixing
+
+import (
+	"bytes"
+	"fmt"
+	"hash"
+
+	"github.com/decred/dcrd/crypto/blake256"
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+	"github.com/decred/dcrd/dcrec/secp256k1/v4/schnorr"
+)
+
+const tag = "decred-mix-signature"
+
+// Signed is an interface describing a signed mixing message.
+type Signed interface {
+	Pub() []byte
+	Sig() []byte
+	Sid() []byte
+	GetRun() uint32
+	Command() string
+	WriteSignedData(hash.Hash)
+}
+
+// SignMessage creates a signature for the message m and writes the signature
+// into the message.
+func SignMessage(m Signed, priv *secp256k1.PrivateKey) error {
+	sig, err := sign(priv, m)
+	if err != nil {
+		return err
+	}
+	copy(m.Sig(), sig)
+	return nil
+}
+
+// VerifySignedMessage verifies that a signed message carries a valid
+// signature for the represented identity.
+func VerifySignedMessage(m Signed) bool {
+	h := blake256.New()
+	m.WriteSignedData(h)
+	sigHash := h.Sum(nil)
+
+	h.Reset()
+
+	command := m.Command()
+	sid := m.Sid()
+	run := m.GetRun()
+	if len(sid) != 32 {
+		sid = zeroSID
+		run = 0
+	}
+
+	return verify(h, m.Pub(), m.Sig(), sigHash, command, sid, run)
+}
+
+// VerifySignature verifies a message signature from its signature hash and
+// information describing the message type and its place in the protocol.
+// Multiple messages of the same command, sid, and run should not be signed by
+// the same public key, and demonstrating this can be used to prove malicious
+// behavior by sending different versions of messages through the network.
+func VerifySignature(pub, sig, sigHash []byte, command string, sid []byte, run uint32) bool {
+	h := blake256.New()
+	return verify(h, pub, sig, sigHash, command, sid, run)
+}
+
+var zeroSID = make([]byte, 32)
```

Comment:
Probably should be `var zeroSID [32]byte` to avoid a guaranteed alloc that the GC has to walk when it may never actually be used.

Reply:
this gets optimized. there's no difference.

Reply:
Once all of this lands, I'll be doing some profiling to confirm these types of things, but I'm positive that historically this was definitely not optimized out.  It's entirely possible the compiler changed in recent times, because I've avoided this pattern for a long time due to previously discovering it wasn't optimized out through profiling a long time ago.

Reply:
For reference:

```sh
$ go version go1.22.3 linux/amd64
$ go build -gcflags="-m" 2>&1 | grep "signatures.go:70"
./signatures.go:70:19: make([]byte, 32) escapes to heap
```

But with `var zeroSID [32]byte`:

```sh
$ go build -gcflags="-m" 2>&1 | grep "signatures.go:70"
```

Reply:
fascinating. I know this would be optimized as a stack allocation if it were in a function body.  perhaps the optimizer run is not doing  the same for placing this in the .bss